### PR TITLE
Introduce distributed.comm.ucx.environment config slot

### PR DESCRIPTION
**Pull request by wence-**
Can be used for setting arbitrary UCX configuration in addition to the specific high-level options in those cases where one needs fine-grained control over the UCX configuration.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

cc @pentschev, @quasiben

### CODE REVIEW

File: distributed/comm/ucx.py [hunk 1]
```diff
@@ -596,14 +603,25 @@ def _scrub_ucx_config():
...
+    # {"some-name": value} is translated to {"SOME_NAME": value}
+    def normalise(k):
+        return "_".join(map(str.upper, k.split("-")))
+
+    environment_options = {
+        normalise(k): v
+        for k, v in dask.config.get("distributed.comm.ucx.environment", {}).items()
+    }
 
-    # ANY UCX options defined in config will overwrite high level dask.ucx flags
```
Comment (wence-) [inline distributed/comm/ucx.py:601]: Many years ago (#3515) we could sort of do what I'm introducing, but that was removed as part of a cleanup in #4850 (when this comment became out of date but was not updated)

File: distributed/comm/ucx.py [hunk 2]
```diff
@@ -596,14 +603,25 @@ def _scrub_ucx_config():
...
-    valid_ucx_vars = list(get_config().keys())
-    for k, v in options.items():
+    # High level options override general environment ones.
```
Comment (wence-) [inline distributed/comm/ucx.py:618]: We can bikeshed if this is the right way round.

File: distributed/comm/ucx.py [hunk 3]
```diff
@@ -551,17 +551,24 @@ def get_local_address_for(self, loc):
...
 
-def _scrub_ucx_config():
-    """Function to scrub dask config options for valid UCX config options"""
+def _prepare_ucx_config():
+    """Translate dask config options to appropriate UCX config options
+
+    Returns
+    -------
+    dict
+        Options suitable for passing to ``ucp.init``
```
[the remark this answers is not in the document]
Comment (wence-) [reply distributed/comm/ucx.py]: Ugh. Fortunately, if we just modify `os.environ` then we can set things, which is what I've done in my most recent commit.

File: distributed/tests/test_config.py [hunk 4]
```diff
@@ -338,13 +338,14 @@ def test_schema_is_complete():
         schema = yaml.safe_load(f)
 
     skip = {
-        "default-task-durations",
-        "bokeh-application",
-        "environ",
-        "pre-spawn-environ",
+        "distributed.scheduler.default-task-durations",
```
Comment (jrbourbeau) [inline distributed/tests/test_config.py:341]: Why are these changes needed?

File: distributed/comm/ucx.py [hunk 5]
```diff
@@ -596,14 +612,27 @@ def _scrub_ucx_config():
...
 
-    # ANY UCX options defined in config will overwrite high level dask.ucx flags
-    valid_ucx_vars = list(get_config().keys())
-    for k, v in options.items():
+    valid_ucx_vars = set(get_config().keys())
+    for k, v in high_level_options.items():
         if k not in valid_ucx_vars:
             logger.debug(
                 f"Key: {k} with value: {v} not a valid UCX configuration option"
             )
```
Comment (pentschev) [inline distributed/comm/ucx.py:636]: I think we will need to remove this. Unfortunately we can't validate non-UCP options, so if someone tries to set `distributed.comm.ucx.environment.tcp-rx-seg-size` for example, this is likely to throw messages that are not correct. Alternatively, we could keep this and make it more explicit, e.g., `... not a valid UCP configuration ...`, but that may be too confusing for those that do not know much about UCX layers.

**Review by pentschev — CHANGES_REQUESTED** [review]
Thanks @wence- , overall this looks great. I've left a few comments that we probably need to change.

File: distributed/comm/ucx.py (hunk 5 above)
Comment (wence-) [reply distributed/comm/ucx.py:636]: Note that this is only validating `high_level_options` (which can only contain `TLS` and `TLS_SOCKADDR_PRIORITY` by construction). So we won't have anything non-UCP here. We could just remove this because the only way slots get set in `high_level_options` is by this function anyway.

I think this is a leftover from when one could put arbitrary key-value pairs in the ucx config dict (about 2 years ago).

File: distributed/tests/test_config.py (hunk 4 above)
Comment (wence-) [reply distributed/tests/test_config.py:341]: The schema validation in this test traverses the full yaml schema from its root to any leaves. Normally, any `object` property (mapping to a `dict` in python) is required to have `properties` set (indicating which keys are valid in the dict). In some cases (for example the nanny environment) the keys can be arbitrary, so we skip validation. 

Previously this skipping was just done on the leaf key name (disregarding the path to the leaf), so could have resulted in false positive skipping (e.g. previously _any_ key name `environ` would not have had its properties validated even in the case where they existed).

To avoid this, I'm changing the skip list to be specify the fully-qualified key names, and remembering the path to the leaf when traversing.

File: distributed/comm/ucx.py (hunk 5 above)
Comment (pentschev) [reply distributed/comm/ucx.py:636]: Yeah, I actually overlooked the fact this is validating `high_level_options`, but I agree in that case it doesn't make any sense to keep it, so I would just remove this.

File: distributed/comm/ucx.py [hunk 6]
```diff
@@ -596,14 +612,27 @@ def _scrub_ucx_config():
...
+    def normalise(k):
+        return "_".join(map(str.upper, ("UCX", *k.split("-"))))
+
+    environment_options = {
+        key: v
+        for k, v in dask.config.get("distributed.comm.ucx.environment", {}).items()
+        # High-level settings take precedence over low-level
+        # environment configuration, as do settings already in the
+        # environment (set externally)
+        if (key := normalise(k)) not in os.environ and key[4:] not in high_level_options
```
[the remark this answers is not in the document]
Comment (wence-) [reply distributed/comm/ucx.py:628]: I think logging at `warning` level is a good idea. I don't want to raise an exception of any kind because for debug purposes this might well be useful (e.g. one has defaults in a `yaml` file but temporarily wants to override on the commandline with environment variables).

[the remark this answers is not in the document]
Comment (pentschev) [reply distributed/comm/ucx.py:628]: Agreed, warnings should be good.

File: distributed/comm/ucx.py (hunk 5 above)
Comment (wence-) [reply distributed/comm/ucx.py]: Done.

File: distributed/comm/ucx.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (wence-) [reply distributed/comm/ucx.py]: I went for warning when the high level options override the ucx.environment, but only info for os.environ overriding ucx.environment (since this is a more common use case).

File: distributed/comm/ucx.py [hunk 7]
```diff
@@ -89,7 +89,7 @@ def init_once():
         return
 
     # remove/process dask.ucx flags for valid ucx options
-    ucx_config = _scrub_ucx_config()
+    ucx_config, ucx_environment = _prepare_ucx_config()
 
     # We ensure the CUDA context is created before initializing UCX. This can't
     # be safely handled externally because communications in Dask start before
```
Comment (wence-) [inline distributed/comm/ucx.py:95]: I note here that this check is not complete since it will only check the high level options for TLS. Shall I extend to also look in `ucx_environment` and `os.environ`?

Comment (pentschev) [reply distributed/comm/ucx.py:95]: Yes, I think it would be better. Since we're opening the door for users to set arbitrarily any variables now, it may be wise to do so. By the way, it tests for `"cuda_copy" in ucx_config["TLS"]`, but it's also allowed to set `"cuda"` (which would include not only `cuda_copy`, but also `cuda_ipc`), could you change the check for `"cuda" in ucx_config["TLS"]` as well?

File: distributed/comm/ucx.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (pentschev) [reply distributed/comm/ucx.py]: Just to close this conversation, according to https://github.com/openucx/ucx/issues/8649 we can only set UCP and UCT variables via `ucp.init()` in UCX-Py, which is a limitation of `ucp_config_modify`. Therefore, handling via `os.environ` like we are doing already is the better option here.

File: distributed/comm/ucx.py (hunk 7 above)
Comment (wence-) [reply distributed/comm/ucx.py:95]: Is that right? If UCX has cuda support, then `UCX_TLS=all` will need the same handling no? I'll do best effort specific checking.

Comment (pentschev) [reply distributed/comm/ucx.py:95]: > If UCX has cuda support, then UCX_TLS=all will need the same handling no?

Yeah, but the problem is with `UCX_TLS=all` we don't know what really is available, nor if the user will indeed require CUDA. For example, we don't want users of `LocalCluster` (the CPU one) to have a CUDA context being created just because it can. The solution for that, although not optimal, was to allow creating the context if `distributed.comm.ucx.create-cuda-context=True`.

Comment (pentschev) [reply distributed/comm/ucx.py:95]: Also this requirement (of having a CUDA context created before instantiating UCX to ensure correct CUDA<->IB mappings) is supposed to go away with UCX protocol v2, which is currently under development. After that, we will hopefully be able to ditch context creation here.

Comment (wence-) [reply distributed/comm/ucx.py:95]: Added best-effort checking.

**Comment by quasiben** [comment]
Generally looks good.  @pentschev @wence-, now that TLS=ALL generally works what do you think about adding that as a control flag to the standard UCX config here as well ?  So instead of having to set all the config options: nvlink, cuda-copy, infiniband, ... we can (just like with dask-cuda), turn it all on.  Would that be too confusing ?

**Comment by wence-** [comment]
> Generally looks good. @pentschev @wence-, now that TLS=ALL generally works what do you think about adding that as a control flag to the standard UCX config here as well ? So instead of having to set all the config options: nvlink, cuda-copy, infiniband, ... we can (just like with dask-cuda), turn it all on. Would that be too confusing ?

If no UCX_TLS is specified, that implies all, so the current code works I think.

**Comment by pentschev** [comment]
> So instead of having to set all the config options: nvlink, cuda-copy, infiniband, ... we can (just like with dask-cuda), turn it all on. Would that be too confusing ?

The only purpose for those flags today is to have a user-friendly way of explicitly enabling/disabling transports that we generally care about. One could just as well set `distributed.comm.ucx.environment.tls=tcp,cuda_copy,cuda_ipc` for example (equivalent to `enable_tcp_over_ucx=True enable_nvlink=True enable_infiniband=False`). Generally I think we want to make things more widely available to users, but as a developer I wouldn't mind removing them, as I can just set `UCX_TLS` specifically to what I need, but that is unlikely to be true for every user, which would require a more detailed understanding at what each transport does.

**Comment by quasiben** [comment]
I'm rerunning the failed jobs.  Many seemed to have failed `test_connection_made_with_extra_conn_args` and I'm being a bit cautious. I'll plan to merge after CI finishes

**Comment by quasiben** [comment]
Looks like the failures are a known issue:
https://github.com/dask/distributed/issues/7168

**Comment by pentschev** [comment]
Thanks @wence- for working on this, and @quasiben for merging! 😄
